[Fix #4628] Add new NestedPercentLiteral cop

asherkach opened the pull request:
Adds the new `NestedPercentLiteral` cop as discussed in #4628.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] All tests(`rake spec`) are passing.
* [x] The new code doesn't generate RuboCop offenses that are checked by `rake internal_investigation`.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Updated cop documentation with `rake generate_cops_documentation` (required only when you've added a new cop or changed the configuration/documentation of an existing cop).

[1]: http://chris.beams.io/posts/git-commit/

Drenmi commented on lib/rubocop/cop/lint/nested_percent_literal.rb:
```diff
@@ -0,0 +1,51 @@
+        def contains_percent_literals?(node)
+          node.children.any? do |child|
+            literal = child.children.first.to_s
+            REGEXES.any? { |regex| literal.match? regex }
```
We support running RuboCop on old rubies (2.0.0 and above) which means you can't use newer features like `#match?` in the codebase. 🙂

asherkach replied:
Yep, I see from the Travis CI failures. Updating to use `String#match` rather than `String#match?`.

Drenmi commented on lib/rubocop/cop/lint/nested_percent_literal.rb:
```diff
@@ -0,0 +1,51 @@
+        # Returns whether node contains any tokens that match a percent literal
+        # invokation.
+        def contains_percent_literals?(node)
+          node.children.any? do |child|
```
Check the `ArrayNode` node extension. There should be a method to check if a node is a percent literal. Then your check can be something like:

```ruby
node.each_child(:array).any? { |child| child.percent_literal? }
```

asherkach replied:
This doesn't seem to work (at the very least, as written, offenses no longer get emitted), presumably because the AST parses the nested percent literals differently (being tokens to the outer nested percent literal). See the TODO in the last test, I think it is related.

Drenmi replied:
Alas. I forgot the parser does this. Check if there's a `Node#str_content` method though. 🙂

asherkach replied:
There is. But it isn't clear to me how it makes sense to use it. Converting the tokens into strings, joining them, and reparsing them doesn't seem appropriate. Is this what you were thinking?

Drenmi commented on lib/rubocop/cop/lint/nested_percent_literal.rb:
```diff
@@ -0,0 +1,51 @@
+
+        private
+
+        # Returns whether node contains any tokens that match a percent literal
```
This comment is probably less descriptive than the method name itself. 😉 I would ✂️ it.

asherkach replied:
Done.

pocke commented on lib/rubocop/cop/lint/nested_percent_literal.rb:
```diff
@@ -0,0 +1,49 @@
+        def contains_percent_literals?(node)
+          node.children.any? do |child|
+            literal = child.children.first.to_s
+            REGEXES.any? { |regex| literal.match(regex) }
```
This cop  crashes with `%W[\xff]`. 

```bash
$ rubocop --cache false --only NestedPercentLiteral
An error occurred while Lint/NestedPercentLiteral cop was inspecting /tmp/tmp.zzU24B1UWS/test.rb:1:4.
To see the complete backtrace run rubocop -d.

1 error occurred:
An error occurred while Lint/NestedPercentLiteral cop was inspecting /tmp/tmp.zzU24B1UWS/test.rb:1:4.
Errors are usually caused by RuboCop bugs.
Please, report your problems to RuboCop's issue tracker.
Mention the following information in the issue report:
0.49.1 (using Parser 2.4.0.0, running on ruby 2.4.1 x86_64-linux)
Inspecting 1 file
.

1 file inspected, no offenses detected
```


Could you use `scrub` like https://github.com/bbatsov/rubocop/blob/6cc91fc8f0305cff3a3182934cf4b786c7f619dc/lib/rubocop/cop/layout/indent_heredoc.rb#L163

asherkach replied:
Done (next commit).

pocke commented on lib/rubocop/cop/lint/nested_percent_literal.rb:
```diff
@@ -0,0 +1,49 @@
+      class NestedPercentLiteral < Cop
+        include PercentLiteral
+
+        MSG = 'Do not nest percent literals.'.freeze
```
[IMO] I think this message is not good, this seems to mention to style problems. 
Users need more descriptive message.
For example:

https://github.com/bbatsov/rubocop/blob/6cc91fc8f0305cff3a3182934cf4b786c7f619dc/lib/rubocop/cop/lint/percent_string_array.rb#L30-L31

asherkach replied:
Done (next commit).

bbatsov commented on spec/rubocop/cop/lint/nested_percent_literal_spec.rb:
```diff
@@ -0,0 +1,48 @@
+
+  let(:message) { 'Do not nest percent literals.' }
+
+  def expect_offense(source)
```
Why are you defining this instead of using the existing helper?

asherkach replied:
Following the pattern [here](https://github.com/bbatsov/rubocop/blob/master/spec/rubocop/cop/lint/percent_string_array_spec.rb#L11-L16), it seems useful to define a method so as to not repeat myself throughout this file. Most likely, this should be a more-widely scoped method, though that refactoring seems appropriate for another PR.

bbatsov commented on spec/rubocop/cop/lint/nested_percent_literal_spec.rb:
```diff
@@ -0,0 +1,48 @@
+    expect(cop.highlights).to eq([source])
+  end
+
+  context 'when using percent literals' do
```
This context seems rather useless to me.

asherkach replied:
Removed (next commit).

asherkach commented:
Think I've addressed all feedback, excepting this [comment](https://github.com/bbatsov/rubocop/pull/4712#discussion_r138340471). Let me know what you are thinking in that regard.

bbatsov commented:
@asherkach This has to be rebased on top of the current master. Sorry for dropping the ball on you - I've been crazy busy.

asherkach commented:
No worries. PTAL, @bbatsov, my apologies it took so long to get it rebased with tests fixed.
